src/tools/miri/src/shims/unix/linux/sync.rs

```diff
@@ -152,12 +158,16 @@ pub fn futex<'tcx>(
             //
             // Thankfully, preemptions cannot happen inside a Miri shim, so we do not need to
             // do anything special to guarantee fence-load-comparison atomicity.
-            this.atomic_fence(AtomicFenceOrd::SeqCst)?;
-            // Read an `i32` through the pointer, regardless of any wrapper types.
-            // It's not uncommon for `addr` to be passed as another type than `*mut i32`, such as `*const AtomicI32`.
-            let futex_val = this.read_scalar_atomic(&addr, AtomicReadOrd::Relaxed)?.to_i32()?;
-            if val == futex_val {
+            let futex_val = this.read_scalar_atomic(&addr, AtomicReadOrd::SeqCst)?.to_u32()?;
+            if if op == futex_lock_pi {
+                futex_val != 0
+            } else {
+                futex_val == val
+            } {
                 // The value still matches, so we block the thread and make it wait for FUTEX_WAKE.
+                if op == futex_lock_pi {
```

Comment:
Every one of these `if` needs a comment explaining why behavior differs between the two ops.

I am wondering if it wouldn't be better to implement `futex_lock_pi` entirely separately, given that it seems to behave quite differently.

Reply:
After reading through the docs -- yes please make these separate new `match` arms, both for lock and unlock. Add a helper function for the logic that determines `let timeout` to avoid duplication.